_blog/2023-10-25-htmx-with-rust.mdx

```diff
@@ -0,0 +1,265 @@
+```rust
+// src/main.rs
+#[derive(Clone, Serialize, Debug)]
+pub enum MutationKind {
+        Create,
+        Delete
+}
+
+#[derive(Clone, Serialize, Debug)]
+pub struct TodoUpdate {
+        mutation_kind: MutationKind,
+        id: i32
+}
+
+#[shuttle_runtime::main]
+async fn main(#[shuttle_shared_db::Postgres] db: PgPool) -> shuttle_axum::ShuttleAxum {
+    sqlx::migrate!()
+        .run(&db)
+        .await
+        .expect("Looks like something went wrong with migrations :(");
+
+        let (tx, rx) = channel::<TodoUpdate>(10);
+    let state = AppState { db };
+
+    let router = Router::new()
+        .route("/", get(home))
+        .route("/stream", get(stream))
+        .route("/todos", get(fetch_todos).post(create_todo))
+        .route("/todos/:id", delete(delete_todo))
+        // new handler - we will make this later
+        .route("/todos/stream", get(handle_stream))
+        .with_state(state)
+        .layer(Extension(tx))
+        .layer(Extension(rx));
+
+    Ok(router.into())
+}
+```
+
+Now we want to send a message from our `create_todo` and `delete_todo` handlers through our channel - we can add them by including our extension in the function signature, then after a successful SQL transaction we simply send a message to the channel:
+```rust
+// src/main.rs
+async fn create_todo(
+    State(state): State<AppState>,
+    Extension(tx): Extension<TodosStream>,          
+    Form(form): Form<TodoNew>,
+) -> impl IntoResponse {
+    let todo = sqlx::query_as::<_, Todo>(
+        "INSERT INTO TODOS (description) VALUES ($1) RETURNING id, description",
+    )   
+    .bind(form.description)
+    .fetch_one(&state.db)
+    .await
+    .unwrap();
+
+    if let Err(e) = tx.send( TodoUpdate {mutation_kind: MutationKind::Create, id: todo.id }) {
+      eprintln!("Tried to send log of record with ID {} created but something went wrong: {e}", todo.id);
+     }
+     
+    TodoNewTemplate { todo }
+}
+
+async fn delete_todo(
+State(state): State<AppState>, 
+        Path(id): Path<i32>,
+    Extension(tx): Extension<TodosStream>,          
+) -> impl IntoResponse {
+    sqlx::query("DELETE FROM TODOS WHERE ID = $1")
+        .bind(id)
+        .execute(&state.db)
+        .await
+        .unwrap();
+
+    if let Err(e) = tx.send( TodoUpdate {mutation_kind: MutationKind::Delete, id }) {
+      eprintln!("Tried to send log of record with ID {id} created but something went wrong: {e}");
+    }
+    StatusCode::OK
+}
+```
+
+Now we need to implement the stream handler, which we can do by creating a `BroadcastStream` and then mapping our stream to Axum SSE events:
+
+```rust
+pub async fn handle_stream(
+    Extension(tx): Extension<TodosStream>,
+) -> Sse<impl Stream<Item = Result<Event, Infallible>>> {
+    let rx = tx.subscribe();
```

Comment:
Why not extract the rx?

Reply:
Originally it was because I didn't actually have the rx as an extension. I've made some changes so that this isn't necessary now though